# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
%matplotlib inline

In [2]:
os.getcwd()

'C:\\Users\\Manager\\springboard projects\\1588204754_Step_2\\Step 2'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
path = "C:\\Users\\Manager\\springboard projects\\1588204754_Step_2\\Step 2\\data"

In [4]:
print ("The current working directory is %s" % path)

The current working directory is C:\Users\Manager\springboard projects\1588204754_Step_2\Step 2\data


In [5]:
df=pd.read_csv("C:\\Users\\Manager\\springboard projects\\1588204754_Step_2\\Step 2\\data\\step3_output.csv",index_col=0)

In [6]:
df.head()

Name       state  summit_elev  vertical_drop  \
2                Hilltop Ski Area      Alaska         2090            294   
4             Sunrise Park Resort     Arizona        11100           1800   
5   Yosemite Ski & Snowboard Area  California         7800            600   
10               Donner Ski Ranch  California         8012            750   
12                  June Mountain  California        10090           2590   

    trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
2       0        0.0          0          0     0       1  ...   
4       0        0.0          0          1     2       3  ...   
5       0        0.0          0          0     0       1  ...   
10      0        0.0          0          0     0       1  ...   
12      0        0.0          0          2     0       0  ...   

    SkiableTerrain_ac  SnowMaking_ac  daysOpenLastYear  yearsOpen  \
2                30.0      30.000000        150.000000       36.0   
4               800.0      80.000000        115.000000       49.0   
5                88.0     174.873239        110.000000       84.0   
10              505.0      60.000000        163.000000       82.0   
12             1500.0     174.873239        115.103943       58.0   

    averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
2              69.0     30.000000      34.00000              152.0   
4             250.0     74.000000      78.00000              104.0   
5             300.0     47.000000      47.00000              107.0   
10            400.0     75.000000      75.00000              170.0   
12            250.0     57.916957      64.16681              128.0   

    NightSkiing_ac  cluster  
2             30.0        0  
4             80.0        1  
5              0.0        1  
10             0.0        1  
12             0.0        1  

[5 rows x 26 columns]

In [7]:
df.shape

(167, 26)

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [8]:
dfo=df.state
dummies=pd.get_dummies(dfo)
df1=pd.concat([df,dummies],axis=1)
df1.head()


Name       state  summit_elev  vertical_drop  \
2                Hilltop Ski Area      Alaska         2090            294   
4             Sunrise Park Resort     Arizona        11100           1800   
5   Yosemite Ski & Snowboard Area  California         7800            600   
10               Donner Ski Ranch  California         8012            750   
12                  June Mountain  California        10090           2590   

    trams  fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
2       0        0.0          0          0     0       1  ...             0   
4       0        0.0          0          1     2       3  ...             0   
5       0        0.0          0          0     0       1  ...             0   
10      0        0.0          0          0     0       1  ...             0   
12      0        0.0          0          2     0       0  ...             0   

    South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
2              0          0     0        0         0           0   
4              0          0     0        0         0           0   
5              0          0     0        0         0           0   
10             0          0     0        0         0           0   
12             0          0     0        0         0           0   

    West Virginia  Wisconsin  Wyoming  
2               0          0        0  
4               0          0        0  
5               0          0        0  
10              0          0        0  
12              0          0        0  

[5 rows x 61 columns]

In [9]:
df1.shape

(167, 61)

In [10]:
df1.state.unique()

array(['Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut',
       'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Missouri', 'Montana',
       'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

In [11]:
df1=df1.drop(['state'],axis=1)

## Standardize the magnitude of numeric features

In [12]:
df1.head()

Name  summit_elev  vertical_drop  trams  \
2                Hilltop Ski Area         2090            294      0   
4             Sunrise Park Resort        11100           1800      0   
5   Yosemite Ski & Snowboard Area         7800            600      0   
10               Donner Ski Ranch         8012            750      0   
12                  June Mountain        10090           2590      0   

    fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
2         0.0          0          0     0       1       0  ...             0   
4         0.0          0          1     2       3       1  ...             0   
5         0.0          0          0     0       1       3  ...             0   
10        0.0          0          0     0       1       5  ...             0   
12        0.0          0          2     0       0       4  ...             0   

    South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
2              0          0     0        0         0           0   
4              0          0     0        0         0           0   
5              0          0     0        0         0           0   
10             0          0     0        0         0           0   
12             0          0     0        0         0           0   

    West Virginia  Wisconsin  Wyoming  
2               0          0        0  
4               0          0        0  
5               0          0        0  
10              0          0        0  
12              0          0        0  

[5 rows x 60 columns]

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [13]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler =preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [15]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [16]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [17]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [18]:
print(y_test)

[47.         49.         68.         59.         59.         71.
 58.         55.         54.         65.         62.         48.
 42.         42.         44.         65.         39.         69.
 79.         64.16681004 55.         45.         79.         64.16681004
 64.         45.         59.         47.         75.         60.
 39.         45.         68.         64.16681004 57.         40.
 72.         34.         35.         92.         59.         46.        ]


In [19]:
type(y_test)

numpy.ndarray

In [20]:
y_test.ndim

1

In [21]:
y_test.shape

(42,)

In [22]:
print(y_pred)

[5.12959194e+01 5.11477566e+01 5.08562222e+01 6.43099270e+01
 5.71340542e+01 6.79745693e+01 5.50999355e+01 4.16761684e+01
 4.07237453e+01 6.82192593e+01 5.57493191e+01 4.53372707e+01
 3.51728420e+01 5.13809414e+01 4.65086880e+01 6.02096462e+01
 4.03215542e+01 7.04365139e+01 7.34808559e+01 6.41132867e+11
 5.72483730e+01 4.43626308e+01 6.41127875e+11 6.41132848e+11
 6.31915493e+01 6.01447659e+01 5.67721767e+01 4.82085171e+01
 6.80873623e+01 6.79961147e+01 3.18477993e+01 5.73295498e+01
 6.56463223e+01 7.10627346e+01 6.16873379e+01 3.89484463e+01
 7.28048306e+01 3.29748440e+01 5.39913235e+01 8.03610744e+01
 5.62337856e+01 4.08380032e+01]


In [23]:
type(y_pred)

numpy.ndarray

In [24]:
y_pred.ndim

1

In [25]:

y_pred.shape

(42,)

In [26]:
import sklearn
sklearn.metrics.explained_variance_score(y_test,y_pred)

-1.5931426535953095e+20

In [27]:
sklearn.metrics.mean_absolute_error(y_test,y_pred)

45795085489.85598

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [28]:
lm.intercept_

11517326944.641867

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [29]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        2.205866e+00
vertical_drop      6.480054e-01
trams              1.018295e+12
fastEight          2.240217e+12
fastSixes          1.332293e+12
fastQuads          1.836577e+11
quad               2.309500e+11
triple             3.819370e+11
double             4.475405e+11
surface            4.624627e+11
total_chairs       8.095307e+11
Runs               2.085247e+00
TerrainParks       3.555908e-01
LongestRun_mi      7.455139e-01
SkiableTerrain_ac  5.057678e-01
SnowMaking_ac      5.603027e-02
daysOpenLastYear   4.916992e-01
yearsOpen          4.987183e-01
averageSnowfall    1.892090e-01
AdultWeekday       1.027338e+01
projectedDaysOpen  3.939514e-01
NightSkiing_ac     1.243134e+00
cluster            9.420776e-01
Alaska             4.946359e+10
Arizona            4.946359e+10
California         1.369226e+11
Colorado           1.092632e+11
Connecticut        1.092632e+11
Idaho              1.193219e+11
Illinois           8.515578e+10
Indiana            6.974106e+10
Iowa               8.515578e+10
Maine              1.092632e+11
Maryland           3.837114e+05
Massachusetts      9.802918e+10
Michigan           1.833154e+11
Minnesota          1.369226e+11
Missouri           6.974106e+10
Montana            8.515578e+10
Nevada             4.946359e+10
New Hampshire      1.590341e+11
New Jersey         1.448897e+03
New Mexico         1.193219e+11
New York           2.035819e+11
North Carolina     8.515578e+10
Ohio               9.802918e+10
Oregon             8.515578e+10
Pennsylvania       1.521183e+11
Rhode Island       5.905151e-03
South Dakota       4.946359e+10
Tennessee          4.946359e+10
Utah               8.515578e+10
Vermont            1.284815e+11
Virginia           6.974106e+10
Washington         9.802918e+10
West Virginia      8.515578e+10
Wisconsin          1.521183e+11
Wyoming            1.092632e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [30]:
df.head(2)

Name    state  summit_elev  vertical_drop  trams  fastEight  \
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  SnowMaking_ac  \
2          0          0     0       1  ...               30.0           30.0   
4          0          1     2       3  ...              800.0           80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
2             150.0       36.0             69.0          30.0          34.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  cluster  
2              152.0            30.0        0  
4              104.0            80.0        1  

[2 rows x 26 columns]

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 167 entries, 2 to 329
Data columns (total 26 columns):
Name                 167 non-null object
state                167 non-null object
summit_elev          167 non-null int64
vertical_drop        167 non-null int64
trams                167 non-null int64
fastEight            167 non-null float64
fastSixes            167 non-null int64
fastQuads            167 non-null int64
quad                 167 non-null int64
triple               167 non-null int64
double               167 non-null int64
surface              167 non-null int64
total_chairs         167 non-null int64
Runs                 167 non-null float64
TerrainParks         167 non-null float64
LongestRun_mi        167 non-null float64
SkiableTerrain_ac    167 non-null float64
SnowMaking_ac        167 non-null float64
daysOpenLastYear     167 non-null float64
yearsOpen            167 non-null float64
averageSnowfall      167 non-null float64
AdultWeekday         167 non-null f

In [35]:
from sklearn import preprocessing
X = df.drop(['Name','state','AdultWeekend'], axis=1)
y = df.AdultWeekend
#scaler =preprocessing.StandardScaler().fit(X)
#X_scaled=scaler.transform(X)

In [36]:
scaler =preprocessing.StandardScaler().fit(X)

In [37]:
X_scaled=scaler.transform(X)

In [38]:
from sklearn.model_selection import train_test_split
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [39]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [40]:
y_pred=lm.predict(X_test)


In [41]:
import sklearn
sklearn.metrics.explained_variance_score(y_test,y_pred)

0.5559565165902212

In [42]:
sklearn.metrics.mean_absolute_error(y_test,y_pred)


6.85299068624076

In [43]:
lm.intercept_

56.241296374899306

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [45]:
from sklearn import preprocessing
X = df.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y = df.AdultWeekend
scaler =preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)

In [46]:
from sklearn.model_selection import train_test_split
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [47]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [48]:
y_pred=lm.predict(X_test)

In [49]:
import sklearn
sklearn.metrics.explained_variance_score(y_test,y_pred)

0.5413080297878095

In [50]:
sklearn.metrics.mean_absolute_error(y_test,y_pred)

7.275361831392762

In [51]:
lm.intercept_

56.16971676283306

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.5931426535953095e+20 |45795085489.85598  |Name,base_elev|
| Model 2. | 0.5559565165902212|6.85299068624076 |'state',Name,base_elev|
| Model 3. | 0.5413080297878095 |7.275361831392762 |'Name,'state','summit_elev','base_elev'|

Model Selection:

Both model 2 and model 3 provide similar kind of performance.i will go with model 2 because it has low MAE in compariosn two model 3